src/games/strategy/engine/framework/startup/mc/GameSelectorModel.java

```diff
@@ -250,20 +233,18 @@ public synchronized ClientModel getClientModelForHostBots() {
   }
 
   /**
-   * We dont have a gane data (ie we are a remote player and the data has not been sent yet), but
-   * we still want to display game info
+   * We dont have a gane data (ie we are a remote player and the data has not been sent yet), but we
+   * still want to display game info
    */
   public void clearDataButKeepGameInfo(final String gameName, final String gameRound, final String gameVersion) {
-    synchronized (this) {
-      m_data = null;
-      m_gameName = gameName;
-      m_gameRound = gameRound;
-      m_gameVersion = gameVersion;
-    }
+    m_data = null;
+    m_gameName = gameName;
+    m_gameRound = gameRound;
+    m_gameVersion = gameVersion;
     notifyObs();
   }
 
-  public synchronized String getFileName() {
```

Comment:
Here's an example of bad synchronization. We synchronize on the wrong object (this) and do not hold the lock for long enough. 

Let's say thread A comes in and calls getFileName(). Thread B comes in and is updating m_data. With this synchronize block we will correctly get the correct value back from getFileName().

But, getting the correct value back from the method is not actually the thread safety problem that we need to solve. We need to be sure that we *use* the correct value of getFileName().

So in this case, thread B will come along and almost immediately change m_data once getFileName() returns and releases the lock. After that point, anything we do with getFileName() will be incorrect since it will not reflect the actual value of m_data at that time.

The way to solve this is have a shared lock around this method and m_data. Clients of this code would need to grab that lock, get this data, finish using it, then release the lock.

To solve this correctly is quite a pain. To boot, actually getting into those scenarios is very hard. Much of this behavior is driven by swing threads, either remote or local. Since the remote one is bound by a human and swing threads, we'll never see thousands of thread hitting this code, chances are good usually it will always be 1, worst case sometimes 2 (and hitting those 2 threads at the same time on this specific code will likely be a once in a lifetime kind of event - though for other larger pieces of code if they existed could perhaps be more vulnerable)